File: exercises/binary_tree_how_high/test_cases.php
<?php

require_once __DIR__ . "/../binary_tree.php";

// --------------------
// test_00
// --------------------
$a0 = new Node('a');
$b0 = new Node('b');
$c0 = new Node('c');
$d0 = new Node('d');
$e0 = new Node('e');
$f0 = new Node('f');

$a0->left = $b0;
$a0->right = $c0;
$b0->left = $d0;
$b0->right = $e0;
$c0->right = $f0;

// expected: 2

// --------------------
// test_01
// --------------------
$a1 = new Node('a');
$b1 = new Node('b');
$c1 = new Node('c');
$d1 = new Node('d');
$e1 = new Node('e');
$f1 = new Node('f');
$g1 = new Node('g');

$a1->left = $b1;
$a1->right = $c1;
$b1->left = $d1;
$b1->right = $e1;
$c1->right = $f1;
$e1->left = $g1;

// expected: 3

// --------------------
// test_02
// --------------------
$a2 = new Node('a');
$c2 = new Node('c');

$a2->right = $c2;

// expected: 1

// --------------------
// test_03
// --------------------
$a3 = new Node('a');

// expected: 0

// --------------------
// test_04
// --------------------
// null root
// expected: -1


// --------------------
// Edge cases
// --------------------

// edge_00: left-skewed chain (height = edges)
$e0a = new Node('a');
$e0b = new Node('b');
$e0c = new Node('c');
$e0d = new Node('d');
$e0a->left = $e0b;
$e0b->left = $e0c;
$e0c->left = $e0d;
// expected: 3

// edge_01: right-skewed chain
$e1a = new Node('a');
$e1b = new Node('b');
$e1c = new Node('c');
$e1a->right = $e1b;
$e1b->right = $e1c;
// expected: 2

// edge_02: unbalanced where deepest path is on one side
$e2a = new Node('a');
$e2b = new Node('b');
$e2c = new Node('c');
$e2d = new Node('d');
$e2e = new Node('e');
$e2a->left = $e2b;
$e2b->left = $e2c;
$e2a->right = $e2d;
$e2d->right = $e2e;
// deepest path a->b->c is height 2, a->d->e is height 2
// expected: 2

return $testCases = [
    "test_00" => [
        "input" => [$a0],
        "expected" => 2,
    ],
    "test_01" => [
        "input" => [$a1],
        "expected" => 3,
    ],
    "test_02" => [
        "input" => [$a2],
        "expected" => 1,
    ],
    "test_03" => [
        "input" => [$a3],
        "expected" => 0,
    ],
    "test_04" => [
        "input" => [null],
        "expected" => -1,
    ],

    // --------------------
    // Edge cases
    // --------------------
    "edge_00" => [
        "input" => [$e0a],
        "expected" => 3,
    ],
    "edge_01" => [
        "input" => [$e1a],
        "expected" => 2,
    ],
    "edge_02" => [
        "input" => [$e2a],
        "expected" => 2,
    ],
];

File: exercises/binary_tree_l2_all_tree_paths/test_cases.py
from binary_tree import Node

# --------------------
# test_00
# --------------------
a0 = Node("a")
b0 = Node("b")
c0 = Node("c")
d0 = Node("d")
e0 = Node("e")
f0 = Node("f")

a0.left = b0
a0.right = c0
b0.left = d0
b0.right = e0
c0.right = f0


# --------------------
# test_01
# --------------------
a1 = Node("a")
b1 = Node("b")
c1 = Node("c")
d1 = Node("d")
e1 = Node("e")
f1 = Node("f")
g1 = Node("g")
h1 = Node("h")
i1 = Node("i")

a1.left = b1
a1.right = c1
b1.left = d1
b1.right = e1
c1.right = f1
e1.left = g1
e1.right = h1
f1.left = i1


# --------------------
# test_02
# --------------------
q2 = Node("q")
r2 = Node("r")
s2 = Node("s")
t2 = Node("t")
u2 = Node("u")
v2 = Node("v")

q2.left = r2
q2.right = s2
r2.right = t2
t2.left = u2
u2.right = v2


# --------------------
# test_03
# --------------------
z3 = Node("z")


# --------------------
# Edge cases
# --------------------

# edge_00: left-skewed tree
el0a = Node(1)
el0b = Node(2)
el0c = Node(3)
el0d = Node(4)
el0a.left = el0b
el0b.left = el0c
el0c.left = el0d
# only one path: [1,2,3,4]

# edge_01: right-skewed tree
er1a = Node(1)
er1b = Node(2)
er1c = Node(3)
er1a.right = er1b
er1b.right = er1c
# only one path: [1,2,3]

# edge_02: root has two children, both leaves
et2a = Node("root")
et2b = Node("left")
et2c = Node("right")
et2a.left = et2b
et2a.right = et2c
# paths: ["root","left"], ["root","right"]

# edge_03: deeper on one side only
ed3a = Node(10)
ed3b = Node(20)
ed3c = Node(30)
ed3d = Node(40)
ed3a.right = ed3b
ed3b.left = ed3c
ed3c.left = ed3d
# only one path: [10,20,30,40]

# edge_04: mixed values
from typing import Any

em4a: Node[Any] = Node("x")
em4b: Node[Any] = Node(1)
em4c: Node[Any] = Node(True)
em4a.left = em4b
em4a.right = em4c
# paths: ["x",1], ["x",True]


test_cases = {
    "test_00": {
        "input": [a0],
        "expected": [
            ["a", "b", "d"],
            ["a", "b", "e"],
            ["a", "c", "f"],
        ],
    },
    "test_01": {
        "input": [a1],
        "expected": [
            ["a", "b", "d"],
            ["a", "b", "e", "g"],
            ["a", "b", "e", "h"],
            ["a", "c", "f", "i"],
        ],
    },
    "test_02": {
        "input": [q2],
        "expected": [
            ["q", "r", "t", "u", "v"],
            ["q", "s"],
        ],
    },
    "test_03": {
        "input": [z3],
        "expected": [
            ["z"],
        ],
    },

    # --------------------
    # Edge cases
    # --------------------
    "edge_00": {
        "input": [el0a],
        "expected": [
            [1, 2, 3, 4],
        ],
    },
    "edge_01": {
        "input": [er1a],
        "expected": [
            [1, 2, 3],
        ],
    },
    "edge_02": {
        "input": [et2a],
        "expected": [
            ["root", "left"],
            ["root", "right"],
        ],
    },
    "edge_03": {
        "input": [ed3a],
        "expected": [
            [10, 20, 30, 40],
        ],
    },
    "edge_04": {
        "input": [em4a],
        "expected": [
            ["x", 1],
            ["x", True],
        ],
    },
}